File: 1093-recover-a-tree-from-preorder-traversal/recover-a-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int last,n;
    TreeNode solve(String t,int d){
        if(last>=n) return null;
        int val=0;
        int i=last,depth=0;
        while(i<n && t.charAt(i)=='-'){
            depth++;i++;
        } 
        if(depth<=d) return null;

        while(i<n && Character.isDigit(t.charAt(i))){
            int v=(int)t.charAt(i)-48;
            val*=10;
            val+=v;
            i++;
        }
        TreeNode node=new TreeNode(val);
        last=i;
        node.left=solve(t,d+1);
        node.right=solve(t,d+1);
        return node;
    }
    public TreeNode recoverFromPreorder(String t) {
        this.n=t.length();
        last=0;
        return solve(t,-1);
    }
}

File: 1093-recover-a-tree-from-preorder-traversal/recover-a-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int last=0;
    TreeNode* solve(string& s,int i,int d){
        if(last>=s.size())
            return nullptr;
        int cur_d=0;
        while(i<s.size() && s[i]=='-'){
            cur_d++;i++;
        }
        if(cur_d<=d)
            return nullptr;
        int val=0;
        while(i<s.size() && isdigit(s[i])){
            int k=s[i]-'0';
            val=val*10+k;
            i++;
        }
        last=i;
        TreeNode* node=new TreeNode(val);
        node->left=solve(s,last,d+1);
        node->right=solve(s,last,d+1);
        return node;
    }
    TreeNode* recoverFromPreorder(string traversal) {
        return solve(traversal,0,-1);
    }
};